src/redux/requests.ts

```diff
@@ -0,0 +1,279 @@
+   */
+  displayDetails: RequestDisplayDetails | null | {};
+
+  /**
+   * The image URL for the dapp, or undefined.
+   */
+  imageUrl: string | undefined;
+
+  /**
+   * The payload for the request.
+   */
+  payload: any;
+}
+
+/**
+ * Display details loaded for a request.
+ */
+interface RequestDisplayDetails {
+  /**
+   * Data loaded for the request, depending on the type of request.
+   */
+  request: any;
+
+  /**
+   * The timestamp for the request.
+   */
+  timestampInMs: number;
+}
+
+/**
+ * Represents the current state of the `requests` reducer.
+ */
+interface RequestsState {
+  /**
+   * Current requests, as an object mapping request IDs to `RequestData`
+   * objects.
+   */
+  requests: {
+    [requestId: string]: RequestData;
+  };
+}
+
+/**
+ * An action for the `requests` reducer.
+ */
+type RequestsAction =
+  | RequestsUpdateRequestsToApproveAction
+  | RequestsClearStateAction;
+
+/**
+ * The action for updating the requests to approve in state.
+ */
+interface RequestsUpdateRequestsToApproveAction {
+  type: typeof REQUESTS_UPDATE_REQUESTS_TO_APPROVE;
+  payload: RequestsState['requests'];
+}
+
+/**
+ * The action for resetting the state.
+ */
+interface RequestsClearStateAction {
+  type: typeof REQUESTS_CLEAR_STATE;
+}
+
+/**
+ * Loads requests from local storage into state.
+ */
+export const requestsLoadState = () => async (
+  dispatch: Dispatch<RequestsUpdateRequestsToApproveAction>,
+  getState: AppGetState
+) => {
+  const { accountAddress, network } = getState().settings;
+  try {
+    const requests = await getLocalRequests(accountAddress, network);
+    const _requests = requests || {};
+    dispatch({ payload: _requests, type: REQUESTS_UPDATE_REQUESTS_TO_APPROVE });
+    // eslint-disable-next-line no-empty
+  } catch (error) {}
+};
+
+/**
+ * Adds a new request to state and updates local storage.
+ *
+ * @param clientId The WalletConnect client ID.
+ * @param peerId The WalletConnect peer ID.
+ * @param requestId The WalletConnect request ID.
+ * @param payload The request payload.
+ * @param peerMeta The WalletConnect peer metadata.
+ */
+export const addRequestToApprove = (
+  clientId: string,
+  peerId: string,
+  requestId: number,
+  payload: any,
+  peerMeta:
+    | undefined
+    | {
+        name?: string;
+        url?: string;
+        scheme?: string;
+      }
+) => (
+  dispatch: Dispatch<RequestsUpdateRequestsToApproveAction>,
+  getState: AppGetState
+) => {
+  const { requests } = getState().requests;
+  const { walletConnectors } = getState().walletconnect;
+  const { accountAddress, network, nativeCurrency } = getState().settings;
+  const walletConnector = walletConnectors[peerId];
+  const chainId = walletConnector._chainId;
+  const dappNetwork = ethereumUtils.getNetworkFromChainId(Number(chainId));
+  const displayDetails = getRequestDisplayDetails(
+    payload,
+    nativeCurrency,
+    dappNetwork
+  );
+  const oneHourAgoTs = Date.now() - EXPIRATION_THRESHOLD_IN_MS;
+  // @ts-ignore
```

Comment:
Why is this needed?

Would also be good to use `@ts-expect-error` instead so it gets cleaned up if it ever becomes redundant.

Reply:
I think the issue here is that `getRequestDisplayDetails` gets interpreted as returning `{} | null`, but it's expected to return the object with `timestampInMs`. There are a series of if statements inside that function, and it does seem to return the expected `timestampInMs` unless the payload doesn't match any of the possible options, but that could go wrong if the payload is invalid somehow. 

I think the possible options would be:
1. Use `@ts-expect-error` (I changed the ignore to this)
2. Cast to `as RequestDisplayDetails`.
3. Update the if statement below to check `displayDetails?.timestampInMs < oneHourAgoTs` (this one probably makes the most sense, unless there's some behavior I'm not realizing where this is supposed to throw an exception if the result is `null`?)

Let me know what you think, thanks!

Reply:
It's tricky when we're avoiding any changes to behaviour. With the current implementation, it would throw an error if `displayDetails` was `null`, whereas changing it to `displayDetails?.timestampInMs < oneHourAgoTs` would make the code more forgiving and potentially open up code paths that were impossible before.

Probably safer to just stick with a `@ts-expect-error` for now, but it might be good to add a comment explaining why it's needed which could help with any future clean up.